### PR TITLE
Added ability to run linux workflows on large runners

### PR DESCRIPTION
**Pull request by rashidnhm**
**Context:**

Currently the CI gets congested when large amounts of pull requests are being updated simultaneously. This pull request gives PRs an escape hatch and use large runners and use different queue to have CI jobs be picked up.

**Description of the Change:**

This pull request adds two new features:
- Ability to add the `urgent` label to any pull request and switch it over to large runners
- Automatic swap of rc branch to large runner
    - This assumes the rc branch is of the format `vX.Y.Z-rcN`

Large runners, albeit slightly more powerful than standard runners, can be spawned at a much higher volume than standard runners ... this is because we pay per minute for these runners vs being included on our GitHub Plan.

If a PR needs CI run without waiting for a runner, **add the `urgent` label to the pull request**.

Important Note:
- This only affect jobs that run on `pull_request` and use `ubuntu` runners.
- This change is already in-place in lightning and catalyst.
    - PennyLaneAI/pennylane-lightning#774
    - PennyLaneAI/catalyst#846

**Benefits:**
Ability to leverage large runner to have quick time for a runner to pick up a job.

**Possible Drawbacks:**
Though we dictate the pool size of large runners, it is possible to still saturate it. 

**Related GitHub Issues:**
None. [sc-73711](https://app.shortcut.com/xanaduai/story/73711/update-pennylane-ci-to-use-large-runner-group)

### CODE REVIEW

File: .github/workflows/determine-workflow-runner.yml
Comment (mudit2812) [inline .github/workflows/determine-workflow-runner.yml:41]:
```suggestion
          RC_BRANCH_FORMAT_REGEX: v[0-9]+\.[0-9]+\.[0-9]+\-rc[0-9]?
```

We usually use dashes between the version number and "rc0", not underscores

File: .github/workflows/format.yml
Comment (mudit2812) [inline .github/workflows/format.yml:20]: I'm not sure it make sense to _potentially_ use the large runners for running `black` and `pylint`?

File: .github/workflows/tests.yml
Comment (mudit2812) [inline .github/workflows/tests.yml:53]: This step should only be run on schedule. I think we missed that when we merged #6164. Since it only runs on schedule, it might not be needed to run this step on the large runners. In that case, we don't need the `determine-runner` step in this file.

```suggestion
    needs: [tests, determine_runner]
    runs-on: ${{ needs.determine_runner.outputs.runner_group }}
    if: ${{ github.event_name == 'schedule' }}

```

**Review by mudit2812 — COMMENTED** [review]
Looks quite good! Just a couple of questions about where it might make sense to _not_ use large runners.

File: .github/workflows/tests.yml
Comment (rashidnhm) [reply .github/workflows/tests.yml:53]: You are right, I will fix it up

**Comment by rashidnhm** [comment]
There is also a syntax error with the `runs-on` expression, I will fix it up Monday morning!

File: .github/workflows/format.yml
Comment (rashidnhm) [reply .github/workflows/format.yml:20]: Would you say we should exclude format/docs_check/upload to codecov from large runners?

Comment (mudit2812) [reply .github/workflows/format.yml:20]: Excluding `upload-to-codecov` and `format` makes sense to me. I think including `docs` has some merit, considering that it's actually very useful for new features or big refactors to verify that the sphinx build passes without issues.

**Review by mudit2812 — COMMENTED** [review]
Are the test runtimes expected to stay the same even while using the large runners? I don't see noticeable improvements in the runtimes based on the most recent CI run.

If that's expected, once the conversation about using large runners with the `docs` and `format` workflows is resolved, I'm happy to approve.

File: .github/workflows/format.yml
Comment (mudit2812) [reply .github/workflows/format.yml:20]: The more I think about it, the more sense it makes to me to use the large runners for the actions above. If a PR is marked as urgent, we should treat all workflows as such.

**Review by mudit2812 — COMMENTED** [review]
Probably a good idea to unconditionally set the `upload.yml` action to use the large runners as well.

**Comment by Alex-Preciado** [comment]
> Automatic swap of rc branch to large runner

Hey @rashidnhm, how big is the large runners pool? I don't know a lot about the specifics but I’m concerned that having all codebases use the large runners pool during feature freeze (when multiple `rc` branches will be created in the ecosystem) might lead to the exact situation we’re trying to avoid—competing for resources. I’m worried this could severely impact Lightning during an already busy period for example.

**Comment by rashidnhm** [comment]
> > Automatic swap of rc branch to large runner
> 
> Hey @rashidnhm, how big is the large runners pool? I don't know a lot about the specifics but I’m concerned that having all codebases use the large runners pool during feature freeze (when multiple `rc` branches will be created in the ecosystem) might lead to the exact situation we’re trying to avoid—competing for resources. I’m worried this could severely impact Lightning during an already busy period for example.

Great question Alex! 

The current pool that is shared across the PL org is 60 runners, this is what GitHub provides as standard for our billing plan. The large runner pool is already set to 150, which is 2.5x the capacity of the standard pool, if we are worried about congestion, we can scale the pool up all the way to **1000**, more than enough for PL needs!

Since we pay per build minute on large runners, there is a lot of flexibility with scaling. 

We can also increase the pool size just during feature freeze and then scale back down after release

**Comment by Alex-Preciado** [comment]
Nice, This is music to my ears 🚀 ... Thank you so much for the details, @rashidnhm !!

File: .github/workflows/determine-workflow-runner.yml
Comment (mudit2812) [inline .github/workflows/determine-workflow-runner.yml:34]: Shouldn't this job get skipped if `force_large_runner` is true?

Comment (mudit2812) [reply .github/workflows/determine-workflow-runner.yml:34]: Oh wait, so the way you've set it up is that we choose the output of `determine_workflow_runner` if it finishes, and if not, only then we care about `force_large_runner.

File: .github/workflows/determine-workflow-runner.yml
Comment (mudit2812) [inline .github/workflows/determine-workflow-runner.yml:51]: I don't understand the purpose of this condition. Why do we want to skip this if the default runner is not `ubuntu-*`?

File: .github/workflows/upload.yml
Comment (mudit2812) [inline .github/workflows/upload.yml:12]: This workflow doesn't have any inputs. `force_large_runner` should always be `true`.

```suggestion
      force_large_runner: true
```

**Review by mudit2812 — COMMENTED** [review]
Thanks Rashid. Just a couple of comments, but very close to approval ready :)

File: .github/workflows/determine-workflow-runner.yml
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:51]: Correct! Currently we only have large runners for Linux (aka: ubuntu-*) ... if a workflow calls this but wants `windows-large`, we return back what they sent and not the large runner group

File: .github/workflows/determine-workflow-runner.yml
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: Great catch! This is a logic bug, going to address it

Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: If force large runner is true, we always want large runner

Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: Oh wait, no logic bug @mudit2812 ... the `determine_workflow_runner` either outputs the name of the large runner group OR doesn't output anything at all.

So it's either the larger runner group from the workflow OR the force large runner logic.

Comment (mudit2812) [reply .github/workflows/determine-workflow-runner.yml:34]: So this is what I understand. Let me know if this is indeed the behaviour we want.
* If the job is used by a pull request, then we always use the output of `determine_workflow_runner` as the source of truth. In this case, we use the large runners iff:
  * The PR has the `urgent` label, or
  * The PR is targeting an RC branch.
* If the job trigger is not a pull request, _only then_ we use `force_large_runner` to determine whether or not to use the large runner or default runner.

File: .github/workflows/upload.yml
Comment (mudit2812) [inline .github/workflows/upload.yml:13]: Just adding whitespace
```suggestion
      force_large_runner: true

  tests:
```

File: .github/workflows/determine-workflow-runner.yml
Comment (mudit2812) [inline .github/workflows/determine-workflow-runner.yml:56]: I just confirmed our branch protection rules with @Alex-Preciado . This regex works perfectly for our RC branch formatting.

Mildly unrelated, but the regex used for the RC branch protection rule is quite general (`*-rc*`). Maybe worth updating that as well to something similar to the expression used here? I've had instances in the past there I couldn't push to private branches because their name was consistent with the RC branch protection rule.

**Review by mudit2812 — APPROVED** [review]
Thanks! There's still one unresolved conversation, but that won't impact how we use the workflow, so approving :)

File: .github/workflows/determine-workflow-runner.yml
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: * If the job is used by a pull request, then we always use the output of `determine_workflow_runner` as the source of truth. In this case, we use the large runners if:
  * The PR has the `urgent` label, or :heavy_check_mark: 
  * The PR is targeting an RC branch. :x: -- The check is running on the rc branch itself!
* If the job trigger is not a pull request, _only then_ we use `force_large_runner` to determine whether or not to use the large runner or default runner. This is because if the workflow trigger is not a pull request, then the `determine_workflow_runner` job will get skipped. :heavy_check_mark:

File: .github/workflows/determine-workflow-runner.yml
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:56]: It's a good idea! The branch protection doesn't exactly allow pure regex, but anything that `fnmatch` takes.. [sample syntax here](https://www.geeksforgeeks.org/fnmatch-unix-filename-pattern-matching-python/)

File: .github/workflows/determine-workflow-runner.yml
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: If a PR is going to be *merged into* rc branch, it would need the `urgent` label, however the checks that'd on the rc branch once the PR is merged would automatically use the large runners

Comment (mudit2812) [reply .github/workflows/determine-workflow-runner.yml:34]: Correct me if I'm wrong, but the condition used below to use the large runners below is
```yaml
 if [[ '${{ contains(github.event.pull_request.labels.*.name, 'urgent') }}' == 'true' || ('${{ github.event.pull_request.head.repo.full_name }}' == "$REPO_FULL_NAME" && '${{ github.event.pull_request.head.ref }}' =~ $RC_BRANCH_FORMAT_REGEX) ]]; then
```
Specifically, for the following part
```yaml
('${{ github.event.pull_request.head.repo.full_name }}' == "$REPO_FULL_NAME" && '${{ github.event.pull_request.head.ref }}' =~ $RC_BRANCH_FORMAT_REGEX)
```
Isn't this saying "if the event is a pull request and the head of the PR is the RC branch, then use the large runner"?

But what you're saying we want is: "if the event is a _push_ into the RC branch, then use the large runner".

So shouldn't the condition be the following?
```yaml
('${{ github.event.push.repo.full_name }}' == "$REPO_FULL_NAME" && '${{ github.event.push.ref }}' =~ $RC_BRANCH_FORMAT_REGEX)
```

Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:34]: Great point @mudit2812 

I have updated the logic so that if a PR has the RC branch as the base, it will automatically use large runner without label;

File: .github/workflows/upload.yml
Comment (PietropaoloFrisoni) [inline .github/workflows/upload.yml:34]: Should the version of all actions be `actions/checkout@v4` and `actions/setup-python@v5`?
I see there are still some outdated actions in older workflows (not touched by this PR)

**Review by PietropaoloFrisoni — APPROVED** [review]
Thank you @rashidnhm, it seems very good to me!

I just have a non-blocking observation, but I don't see any problem at this stage

File: .github/workflows/upload.yml
Comment (rashidnhm) [reply .github/workflows/upload.yml:34]: Yes, it's out of scope of this PR but we should update it across the board (all repos)

**Comment by mudit2812** [comment]
Since @rashidnhm is away today, I'm going to merge this once all checks pass.

**Pull request merged**
